# https://samtools.github.io/hts-specs/SAMv1.pdf 

rule trim_short_reads:
    input:
        "/data/short-reads.fq.gz"
    output:
        "/data/trimmed-short-reads.fasta"
    shell:
        "python3 trim_by_tag_length.py /data/short-reads.fq.gz 10 > /data/trimmed-short-reads.fasta"

rule trim_long_reads:
    input:
        "/data/long-reads.fq.gz"
    output:
        "/data/trimmed-long-reads.fasta"
    script:
        "trim_by_tag_length.py /data/long-reads.fq.gz 92 > /data/trimmed-long-reads.fasta"

rule install_bwa:
    output:
        "bwa-mem2-2.0pre2_x64-linux/bwa-mem2"
    shell:
        "curl -L https://github.com/bwa-mem2/bwa-mem2/releases/download/v2.0pre2/bwa-mem2-2.0pre2_x64-linux.tar.bz2 | tar jxf -"

rule map_short_reads:
    input:
        "bwa-mem2-2.0pre2_x64-linux/bwa-mem2",
        "/data/reference.fasta",
        "/data/trimmed-short-reads.fasta"
    output:
        "/data/mapping.sam"
    shell:
        """
        bwa-mem2-2.0pre2_x64-linux/bwa-mem2 index /data/reference.fasta
        bwa-mem2-2.0pre2_x64-linux/bwa-mem2 mem /data/reference.fasta /data/trimmed-short-reads.fasta > mapping.sam
        """

rule map_long_reads:
    input:
        "/data/reference.fasta",
        "/data/trimmed-long-reads.fasta"
    output:
        "/data/mapping.sam"
    conda:
        "minimap2.yml"
    shell:
        """
        minimap2 -x map-ont -d reference /data/reference.fasta > /dev/null 2>&1
        minimap2 -c -a -o /data/mapping.nonunique.sam -N 1 -x map-ont reference /data/trimmed-long-reads.fasta
        samtools view -bq 1 /data/mapping.nonunique.sam > /data/mapping.sam
        """

rule convert_sam_to_bam:
    input:
        "/data/mapping.sam"
    output:
        "/data/mapping.bam",
        "/data/unmapped.fa"
    conda:
        "samtools.yml"
    shell:
        """
        samtools view -S -b /data/mapping.sam > /data/mapping.bam
        samtools view -f 4 /data/mapping.bam > /data/unmapped.sam
        samtools view -S -b /data/unmapped.sam > /data/unmapped.bam
        samtools bam2fq /data/unmapped.bam | seqtk seq -A - > /data/unmapped.fa
        """

rule create_insertion_plot:
    input:
        "/data/mapping.sam"
    output:
        "/data/summary-stats.tsv"
    script:
        "sam_to_insert_plot.py /data/mapping.sam /data/reference.fasta"
